src/mds/Server.cc

```diff
@@ -6286,12 +6308,11 @@ void Server::handle_remove_vxattr(MDRequestRef& mdr, CInode *cur)
     mdr->no_early_reply = true;
     journal_and_reply(mdr, cur, 0, le, new C_MDS_inode_update_finish(this, mdr, cur));
     return;
-  } else if (name == "ceph.dir.layout.pool_namespace"
-          || name == "ceph.file.layout.pool_namespace") {
-    // Namespace is the only layout field that has a meaningful
-    // null/none value (empty string, means default layout).  Is equivalent
-    // to a setxattr with empty string: pass through the empty payload of
-    // the rmxattr request to do this.
+  } else if (name.find("ceph.") == 0) {
```

Comment:
Looking at the complete change-set again, this check is not actually necessary because this method is only called with vxattrs. In fact, I would like to see this entire method removed. The "ceph.dir.layout" removal procedure above should be consolidated into `handle_set_vxattr` along with the others. `handle_client_removevxattr` should call `handle_set_vxattr` instead of `handle_remove_vxattr`.

Reply:
The block of code in handle_remove_vxattr for `ceph.dir.layout` appears to clear/remove the value. 

Other "ceph.*" xattrs, appear to remove the value by setting a value of -1 or 0. 

IIUC, It seems to me at this point, keeping actual removal in the current form separate from `handle_set_vxattr` is fine.

Reply:
There is a lot of duplicated code and, already, there is some undesirable divergence in behavior. See:

https://github.com/ceph/ceph/blob/785c1083fa93f41d0dcbb7f16a651615bbb44771/src/mds/Server.cc#L6070-L6105

https://github.com/ceph/ceph/blob/785c1083fa93f41d0dcbb7f16a651615bbb44771/src/mds/Server.cc#L6332-L6348

vs.

https://github.com/ceph/ceph/blob/785c1083fa93f41d0dcbb7f16a651615bbb44771/src/mds/Server.cc#L6360-L6392

For example, ctime/change_attr is not updated here:

https://github.com/ceph/ceph/blob/785c1083fa93f41d0dcbb7f16a651615bbb44771/src/mds/Server.cc#L6332-L6333

Also, the root directory cannot have its layout removed (and restored to default)? That doesn't make sense to me as you could just set it to whatever the default is. So ya, I'd like this cleaned up. Please :)

Reply:
Done.